File: NetDeck/ActionsTableViewController.m
//
//  ActionsTableViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 17.05.13.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

#import "ActionsTableViewController.h"
// #import "CardFilterViewController.h"
// #import "BrowserFilterViewController.h"
// #import "SavedDecksList.h"

typedef NS_ENUM(NSInteger, NRMenuItem)
{
    NRMenuDecks,
    NRMenuDeckDiff,
    NRMenuCardBrowser,
    NRMenuSettings,
    NRMenuAbout,
    
    NRMenuItemCount
};

@interface ActionsTableViewController()

@property UINavigationController* navController;
@property SettingsViewController* settings;
@property Card* searchForCard;

@end

@implementation ActionsTableViewController

-(void) dealloc
{
    self.tableView.delegate = nil;
    self.tableView.dataSource = nil;
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

-(void) viewDidLoad
{
    [super viewDidLoad];

    self.tableView.scrollEnabled = NO;
    
    self.navigationController.navigationBar.barTintColor = [UIColor whiteColor];
    
    self.title = @"Net Deck";
    
    UIView* tableFoot = [[UIView alloc] initWithFrame:CGRectZero];
    [self.tableView setTableFooterView:tableFoot];
    
    [self.version setTitle:[AppDelegate appVersion]];
    
    NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(loadDeck:) name:Notifications.loadDeck object:nil];
    [nc addObserver:self selector:@selector(newDeck:) name:Notifications.newDeck object:nil];
    [nc addObserver:self selector:@selector(newDeck:) name:Notifications.browserNew object:nil];
    [nc addObserver:self selector:@selector(importDeckFromClipboard:) name:Notifications.importDeck object:nil];
    [nc addObserver:self selector:@selector(loadCards:) name:Notifications.loadCards object:nil];
    [nc addObserver:self selector:@selector(loadCards:) name:Notifications.dropboxChanged object:nil];
    [nc addObserver:self selector:@selector(listDecks:) name:Notifications.browserFind object:nil];
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    BOOL displayedAlert = [CardUpdateCheck checkCardUpdateAvailable:self];
    
    if (!displayedAlert) {
        [AppUpdateCheck checkUpdate];
    }
    
#if !DEBUG
    // first start with this version?
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
    // [defaults setObject:@"" forKey:LAST_START_VERSION];
    NSString* lastVersion = [defaults stringForKey:SettingsKeys.LAST_START_VERSION];
    NSString* thisVersion = [AppDelegate appVersion];
    if ([thisVersion isEqualToString:lastVersion])
    {
        // yes, first start. show "about" tab
        NSIndexPath* indexPath = [NSIndexPath indexPathForRow:NRMenuAbout inSection:0];
        [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
        [self tableView:self.tableView didSelectRowAtIndexPath:indexPath];
        [defaults setObject:thisVersion forKey:SettingsKeys.LAST_START_VERSION];
        [defaults synchronize];
        return;
    }
#endif
    
    if (![CardManager cardsAvailable] || ![PackManager packsAvailable])
    {
        [self resetDetailView];
        return;
    }
    
    // select Decks view
    NSIndexPath* indexPath = [NSIndexPath indexPathForRow:NRMenuDecks inSection:0];
    [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
    [self tableView:self.tableView didSelectRowAtIndexPath:indexPath];
}

-(void) resetDetailView
{
    EmptyDetailViewController *empty = [[EmptyDetailViewController alloc] initWithNibName:@"EmptyDetailView" bundle:nil];
    
    self.navController = [[UINavigationController alloc] initWithRootViewController:empty];
    
    DetailViewManager *detailViewManager = (DetailViewManager*)self.splitViewController.delegate;
    detailViewManager.detailViewController = self.navController;
}

#pragma mark notifications

-(void)loadDeck:(NSNotification*) notification
{
    NSDictionary* userInfo = notification.userInfo;
    
    NRRole role = [[userInfo objectForKey:@"role"] intValue];
    NSString* filename = [userInfo objectForKey:@"filename"];
    
    CardFilterViewController *filter = [[CardFilterViewController alloc] initWithRole:role andFile:filename];
    
    UINavigationController* nc = self.navigationController;
    [nc pushViewController:filter animated:YES];
}

-(void)newDeck:(NSNotification*) notification
{
    NSDictionary* userInfo = notification.userInfo;
    CardFilterViewController* filter;
    
    if ([notification.name isEqualToString:Notifications.browserNew]) {
        Card* card = [CardManager cardByCode:[userInfo objectForKey:@"code"]];
        Deck* deck = [[Deck alloc] initWithRole:card.role];
        [deck addCard:card copies:1];
        
        filter = [[CardFilterViewController alloc] initWithRole:deck.role andDeck:deck];
    } else {
        NRRole role = [[userInfo objectForKey:@"role"] intValue];
        filter = [[CardFilterViewController alloc] initWithRole:role];
    }
    
    UINavigationController* nc = self.navigationController;
    if (nc.viewControllers.count > 1) {
        [nc popToRootViewControllerAnimated:NO];
    }
    
    [nc pushViewController:filter animated:NO];
}

-(void)importDeckFromClipboard:(NSNotification*) notification
{
    NSDictionary* userInfo = notification.userInfo;
    Deck* deck = [userInfo objectForKey:@"deck"];
    [deck saveToDisk];
    
    CardFilterViewController *filter = [[CardFilterViewController alloc] initWithRole:deck.identity.role andDeck:deck];
    
    UINavigationController* nc = self.navigationController;    
    if (nc.viewControllers.count > 1) {
        [nc popToRootViewControllerAnimated:NO];
    }
    
    [nc pushViewController:filter animated:NO];
}

-(void)loadCards:(id) sender
{
    [self.tableView reloadData];

    id delegate = self.splitViewController.delegate;
    if ([delegate isKindOfClass:[DetailViewManager class]]) {
        DetailViewManager* manager = (DetailViewManager*)delegate;
        UIViewController* detail = manager.detailViewController;
        
        if ([detail.childViewControllers.firstObject isKindOfClass:[EmptyDetailViewController class]]) {
            NSIndexPath* indexPath = [NSIndexPath indexPathForRow:NRMenuDecks inSection:0];
            [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
            [self tableView:self.tableView didSelectRowAtIndexPath:indexPath];
        }
    }
}

-(void) listDecks:(NSNotification*)sender
{
    [self.navigationController popToRootViewControllerAnimated:NO];
    self.searchForCard = [CardManager cardByCode:[sender.userInfo objectForKey:@"code"]];
}

#pragma mark Table view data source

- (NSInteger)tableView:(UITableView *)aTableView numberOfRowsInSection:(NSInteger)section
{
    return NRMenuItemCount;
}

- (UITableViewCell *)tableView:(UITableView *)aTableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *cellIdentifier = @"actions";
    
    // Dequeue or create a cell of the appropriate type.
    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        cell.textLabel.font = [UIFont systemFontOfSize:17];
    }
    
    BOOL cardsAvailable = [CardManager cardsAvailable] && [PackManager packsAvailable];
    // Set appropriate labels for the cells.
    switch (indexPath.row)
    {
        case NRMenuAbout:
            cell.textLabel.text = l10n(@"About");
            break;
        case NRMenuSettings:
            cell.textLabel.text = l10n(@"Settings");
            break;
        case NRMenuDeckDiff:
            cell.textLabel.text = l10n(@"Compare Decks");
            cell.textLabel.enabled = cardsAvailable;
            break;
        case NRMenuDecks:
            cell.textLabel.text = l10n(@"Decks");
            cell.textLabel.enabled = cardsAvailable;
            break;
        case NRMenuCardBrowser:
            cell.textLabel.text = l10n(@"Card Browser");
            cell.textLabel.enabled = cardsAvailable;
            break;
    }
    
    return cell;
}

#pragma mark Table view selection

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell* cell = [tableView cellForRowAtIndexPath:indexPath];
    if (!cell.textLabel.enabled)
    {
        [self resetDetailView];
        return;
    }
    
    self.settings = nil;
    
    // Get a reference to the DetailViewManager.
    // DetailViewManager is the delegate of our split view.
    DetailViewManager *detailViewManager = (DetailViewManager*)self.splitViewController.delegate;

    switch (indexPath.row)
    {
        case NRMenuDecks:
        {
            SavedDecksList* decks;
            if (self.searchForCard == nil)
            {
                decks = [[SavedDecksList alloc] init];
            }
            else
            {
                decks = [[SavedDecksList alloc] initWithCard:self.searchForCard];
                self.searchForCard = nil;
            }
            self.navController = [[UINavigationController alloc] initWithRootViewController:decks];
            detailViewManager.detailViewController = self.navController;
            break;
        }
            
        case NRMenuDeckDiff:
        {
            CompareDecksList* decks = [[CompareDecksList alloc] init];
            self.navController = [[UINavigationController alloc] initWithRootViewController:decks];
            detailViewManager.detailViewController = self.navController;
            break;
        }
            
        case NRMenuCardBrowser:
        {
            UINavigationController* nc = self.navigationController;
            BrowserFilterViewController* browser = [[BrowserFilterViewController alloc] init];
            [nc pushViewController:browser animated:NO];
            break;
        }
            
        case NRMenuSettings:
        {
            self.settings = [[SettingsViewController alloc] init];
            self.navController = [[UINavigationController alloc] initWithRootViewController:self.settings.iask];
            detailViewManager.detailViewController = self.navController;
            break;
        }
            
        case NRMenuAbout:
        {
            AboutViewController* about = [[AboutViewController alloc] init];
            
            self.navController = [[UINavigationController alloc] initWithRootViewController:about];
            detailViewManager.detailViewController = self.navController;
            break;
        }
    }
}

@end

File: NetDeck/ActionsTableViewController.swift
//
//  ActionsTableViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 23.01.17.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import UIKit

enum MenuItem: Int {
    case decks, deckDiff, browser, settings, about
    
    var cardsRequired: Bool {
        switch self {
        case .decks, .deckDiff, .browser: return true
        case .settings, .about: return false
        }
    }
    static var count: Int {
        return 5
    }
}

class ActionsTableViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, StartViewController {

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var version: UIBarButtonItem!
    
    private var searchForCard: Card?
    private var selectedItem = MenuItem.decks

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.isScrollEnabled = false
        
        self.title = "Net Deck"
        
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        
        self.version.title = Utils.appVersion()
        
        let nc = NotificationCenter.default
        nc.addObserver(self, selector: #selector(self.loadDeck(_:)), name: Notifications.loadDeck, object: nil)
        nc.addObserver(self, selector: #selector(self.newDeck(_:)), name: Notifications.newDeck, object: nil)
        nc.addObserver(self, selector: #selector(self.newDeck(_:)), name: Notifications.browserNew, object: nil)
        nc.addObserver(self, selector: #selector(self.importDeckFromClipboard(_:)), name: Notifications.importDeck, object: nil)
        nc.addObserver(self, selector: #selector(self.loadCards(_:)), name: Notifications.loadCards, object: nil)
        nc.addObserver(self, selector: #selector(self.listDecks(_:)), name: Notifications.browserFind, object: nil)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if CardManager.cardsAvailable {
            self.selectDecks()
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        let displayedAlert = CardUpdateCheck.checkCardUpdateAvailable()
        if !displayedAlert {
            AppUpdateCheck.checkUpdate()
        }
        
        if !CardManager.cardsAvailable {
            self.resetDetailView()
        } else {
            self.selectDecks()
        }
    }
    
    func addNewDeck(_ role: Role) {
        let filter = CardFilterViewController(role: role)
   
        if let nav = self.navigationController {
            if nav.viewControllers.count > 1 {
                nav.popToRootViewController(animated: false)
            }
            nav.pushViewController(filter, animated: true)
        }
    }

    func openBrowser() {
        let indexPath = IndexPath(row: MenuItem.browser.rawValue, section: 0)
        
        self.tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
        self.tableView(self.tableView, didSelectRowAt: indexPath)
    }
    
    // select "Decks" view
    private func selectDecks() {
        let indexPath = IndexPath(row: MenuItem.decks.rawValue, section: 0)
        
        self.tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
        self.tableView(self.tableView, didSelectRowAt: indexPath)
    }

    @objc func loadDeck(_ notification: Notification) {
        guard let roleCode = notification.userInfo?["role"] as? Int,
            let role = Role(rawValue: roleCode),
            let filename = notification.userInfo?["filename"] as? String else {
                return
        }
        
        let filter = CardFilterViewController(role: role, andFile: filename)
        self.navigationController?.pushViewController(filter, animated: true)
    }
    
    @objc func newDeck(_ notification: Notification) {
        let filter: CardFilterViewController
        if notification.name == Notifications.browserNew {
            guard
                let code = notification.userInfo?["code"] as? String,
                let card = CardManager.cardBy(code: code)
            else { return }
            
            let deck = Deck(role: card.role)
            deck.addCard(card, copies: 1)
            filter = CardFilterViewController(role: card.role, andDeck: deck)
        } else {
            guard
                let roleCode = notification.userInfo?["role"] as? Int,
                let role = Role(rawValue: roleCode)
            else { return }
            filter = CardFilterViewController(role: role)
        }
        
        if let nav = self.navigationController {
            if nav.viewControllers.count > 1 {
                nav.popToRootViewController(animated: false)
            }
            nav.pushViewController(filter, animated: true)
        }
    }
    
    @objc func importDeckFromClipboard(_ notification: Notification) {
        guard
            let deck = notification.userInfo?["deck"] as? Deck,
            let identity = deck.identity
        else { return }
        
        deck.saveToDisk()
        let filter = CardFilterViewController(role: identity.role, andDeck: deck)
        if let nav = self.navigationController {
            if nav.viewControllers.count > 1 {
                nav.popToRootViewController(animated: false)
            }
            nav.pushViewController(filter, animated: true)
        }
    }
    
    @objc func loadCards(_ notification: Notification) {
        self.tableView.reloadData()        
        if self.selectedItem != .settings {
            self.selectDecks()
        }
    }
    
    @objc func listDecks(_ notification: Notification) {
        guard
            let code = notification.userInfo?["code"] as? String,
            let card = CardManager.cardBy(code: code)
        else { return }
        
        _ = self.navigationController?.popToRootViewController(animated: false)
        self.searchForCard = card
        self.selectDecks()
    }
    
    // MARK: - table view
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return MenuItem.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "actionCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier) ?? {
            let c = UITableViewCell(style: .default, reuseIdentifier: identifier)
            c.accessoryType = .disclosureIndicator
            c.textLabel?.font = UIFont.systemFont(ofSize: 17)
            return c
        }()
        
        let menuItem = MenuItem(rawValue: indexPath.row)!
        
        let label: String
        switch menuItem {
        case .about:
            label = "About".localized()
        case .settings:
            label = "Settings".localized()
        case .deckDiff:
            label = "Compare Decks".localized()
        case .decks:
            label = "Decks".localized()
        case .browser:
            label = "Card Browser".localized()
        }
        cell.textLabel?.text = label
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        self.selectedItem = MenuItem(rawValue: indexPath.row)!
        
        if !CardManager.cardsAvailable && self.selectedItem.cardsRequired {
            self.resetDetailView()
            return
        }
        
        switch self.selectedItem {
        case .decks:
            let decks: SavedDecksList
            if let card = self.searchForCard {
                decks = SavedDecksList(card: card)
                self.searchForCard = nil
            } else {
                decks = SavedDecksList()
            }
            self.showAsDetailViewController(decks)
        case .deckDiff:
            let compare = CompareDecksList()
            self.showAsDetailViewController(compare)
        case .browser:
            let browser = BrowserFilterViewController()
            self.navigationController?.pushViewController(browser, animated: true)
        case .settings:
            let settings = Settings.viewController
            Analytics.logEvent(.showSettings)
            self.showAsDetailViewController(settings)
        case .about:
            let about = AboutViewController()
            self.showAsDetailViewController(about)
        }
    }
    
    private func resetDetailView() {
        let empty = EmptyDetailViewController()
        self.showAsDetailViewController(empty)
    }
    
    private func showAsDetailViewController(_ vc: UIViewController) {
        let navController = UINavigationController(rootViewController: vc)
        navController.navigationBar.barTintColor = .white
        self.showDetailViewController(navController, sender: self)
    }
    
}
